lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	IPList []struct {
+		IP string `json:"ip"`
+	} `json:"ipList"`
+}
+
+// powerFlexSdc represents a SDC in PowerFlex.
+type powerFlexSdc struct {
+	ID       string `json:"id"`
+	Name     string `json:"name"`
+	HostType string `json:"hostType"`
+	NQN      string `json:"nqn"`
+}
+
+// powerFlexVolume represents a volume in PowerFlex.
+type powerFlexVolume struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	VolumeType    string `json:"volumeType"`
+	MappedSdcInfo []struct {
+		SdcID    string `json:"sdcId"`
+		SdcName  string `json:"sdcName"`
+		NQN      string `json:"nqn"`
+		HostType string `json:"hostType"`
+	} `json:"mappedSdcInfo"`
+}
+
+// Error tries to return all kinds of errors from the PowerFlex API in a nicely formatted way.
+func (p *powerFlexError) Error() string {
+	var errorStrings []string
+	for k, v := range *p {
+		errorStrings = append(errorStrings, fmt.Sprintf("%s: %v", k, v))
+	}
+
+	return strings.Join(errorStrings, ", ")
+}
+
+// ErrorCode extracts the errorCode value from a PowerFlex response.
+func (p *powerFlexError) ErrorCode() float64 {
+	code, ok := (*p)["errorCode"].(float64)
+	if !ok {
+		return 0
+	}
+
+	return code
+}
+
+// HTTPStatusCode extracts the httpStatusCode value from a PowerFlex response.
+func (p *powerFlexError) HTTPStatusCode() float64 {
+	code, ok := (*p)["httpStatusCode"].(float64)
+	if !ok {
+		return 0
+	}
+
+	return code
+}
+
+// newTokenFactory creates a new instance of an access token factory.
+func newTokenFactory() *tokenFactory {
+	return &tokenFactory{
+		mutex: &sync.RWMutex{},
+	}
+}
+
+// Get returns the current access token of the factory.
+func (t *tokenFactory) Get() string {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+	return t.token
+}
+
+// Set creates a new access token in the factory.
+// An expiry can be set after which the access token will be removed from the factory.
+func (t *tokenFactory) Set(token string, expiry time.Duration) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	go func() {
+		timer := time.NewTimer(expiry)
+		<-timer.C
+		t.mutex.Lock()
+		defer t.mutex.Unlock()
+		t.token = ""
+	}()
+
+	t.token = token
+}
+
+// newPowerFlexClient creates a new instance of the HTTP PowerFlex client.
+func newPowerFlexClient(driver *powerflex) *powerFlexClient {
+	return &powerFlexClient{
+		driver: driver,
+		token:  newTokenFactory(),
+	}
+}
+
+// createBodyReader creates a reader for the given request body contents.
+func (p *powerFlexClient) createBodyReader(contents map[string]any) (io.Reader, error) {
+	body := &bytes.Buffer{}
+	encoder := json.NewEncoder(body)
+	err := encoder.Encode(contents)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to write request body: %w", err)
+	}
+
+	return body, nil
+}
+
+// request issues a HTTP request against the PowerFlex gateway.
+func (p *powerFlexClient) request(method string, path string, body io.Reader, v any) error {
```

Comment:
Could you update the arg `v` to be clearer what it is without needing to look at the internal implementation. Be kind to your friendly review :)

Reply:
Sure, will go with `response` :)